In [324]:
import numpy as np
from scipy import stats
import openpyxl as px
import statsmodels.api as sm
import pandas as pd

# Úkol 1
Pro výpočty v úkolu 1 je využit $\Chi^2$ - test dobré shody.<br>
Je třeba získat skutečné četnosti z odpovědí respondentů ($f_{real_j}$) a teoretické četnosti ($f_{teor_j}$) pomocí bodového odhadu p:<br>
$p = x / n$, kde x je celkový počet respondentů, kterým vyhovuje zimní čas, a n je celkový počet respondentů.<br>
Jednotlivé teoretické četnosti pak lze získat vynásobením počtu respondentů v dané skupině bodovým odhadem p:<br>
f<sub>teor<sub>j</sub></sub> = p * n<sub>j</sub><br>
Testovací kritérium lze pak vypočítat následujícím způsobem:<br>
$t = \sum_{j=1}^{m} \frac{(f_{real_j} - f_{teor_j})^2}{f_{teor_j}} \overset{as}{\sim} \Chi^2(m-q-1)$, kde m je počet tříd a q je počet odhadovaných parametrů. <br>
Doplněk kritického oboru lze vyjádřit následovně:<br>
$\bar{W}_\alpha = \Biggl \langle 0, \Chi^2_{1-\alpha} \biggr \rangle$, kde $\Chi^2_{1-\alpha}$ je kvantil Pearsonova rozdělení s m-q-1 stupni volnosti.

In [325]:
# 2 velka mesta, 2 mala mesta, 3 obce, 1 okoli studenta
resp = np.array([1327, 915, 681, 587, 284, 176, 215, 34])
zimni = np.array([510, 324, 302, 257, 147, 66, 87, 15])
letni = np.array([352, 284, 185, 178, 87, 58, 65, 8])
stridat = np.array([257, 178, 124, 78, 44, 33, 31, 4])
beznazoru = np.array([208, 129, 70, 74, 6, 19, 32, 7])

resp_sk = np.array([np.sum(resp[:2]), np.sum(resp[2:4]), np.sum(resp[4:7])])
zimni_sk = np.array([np.sum(zimni[:2]), np.sum(zimni[2:4]), np.sum(zimni[4:7])])
letni_sk = np.array([np.sum(letni[:2]), np.sum(letni[2:4]), np.sum(letni[4:7])])
stridat_sk = np.array([np.sum(stridat[:2]), np.sum(stridat[2:4]), np.sum(stridat[4:7])])
beznazoru_sk = np.array([np.sum(beznazoru[:2]), np.sum(beznazoru[2:4]), np.sum(beznazoru[4:7])])


In [326]:
def chisq(respondenti, testovane, odhad, st_volnosti):
    testovane_teor = respondenti * p
    print("Skutečné četnosti:", testovane)
    print("Teoretické četnosti: ", testovane_teor)
    t = np.sum(np.square(testovane - testovane_teor)/testovane_teor)
    print("Testovací kritérium: ", t)
    krit_obor = stats.chi2.ppf(0.95, df=st_volnosti)
    print("Doplněk kritického oboru: < 0,", krit_obor, ">")

<b>a) H<sub>0</sub>: V městech, obcích a v okolí studenta je stejné procentuální zastoupení obyvatel, co preferují zimní čas.</b><br>
H<sub>A</sub>: V městech, obcích a v okolí studenta není stejné procentuální zastoupení obyvatel, co preferují zimní čas.

Při ověřování následujících třech hypotéz ( a), b) a c) ) se pracovalo s 8 skupinami. Hodnota stupně volnosti bude 6, neboť je prováděn bodový odhad (tedy 8-1-1).

In [327]:
p = np.sum(zimni)/np.sum(resp)
chisq(resp, zimni, p, 6)

Skutečné četnosti: [510 324 302 257 147  66  87  15]
Teoretické četnosti:  [537.21640199 370.42427115 275.6928182  237.63830292 114.9732164
  71.25100735  87.03958284  13.76439915]
Testovací kritérium:  20.704110374775837
Doplněk kritického oboru: < 0, 12.591587243743977 >


$20,704 \notin \langle 0; 12,592 \rangle$, tedy H<sub>0</sub> zamítáme.

<b>b) H<sub>0</sub>: V městech, obcích a v okolí studenta je stejné procentuální zastoupení obyvatel, co preferují letní čas.</b><br>
H<sub>A</sub>: V městech, obcích a v okolí studenta není stejné procentuální zastoupení obyvatel, co preferují letní čas.

In [328]:
p = np.sum(letni)/np.sum(resp)
chisq(resp, letni, p, 6)

Skutečné četnosti: [352 284 185 178  87  58  65   8]
Teoretické četnosti:  [382.78241289 263.938137   196.4392036  169.32424745  81.92178241
  50.76842854  62.01825077   9.80753733]
Testovací kritérium:  6.932364791415857
Doplněk kritického oboru: < 0, 12.591587243743977 >


$6,932 \in \langle 0; 12,592 \rangle$, tedy H<sub>0</sub> nezamítáme.

<b>c) H<sub>0</sub>: V městech, obcích a v okolí studenta je stejné procentuální zastoupení obyvatel, co preferují střídání času.</b><br>
H<sub>A</sub>: V městech, obcích a v okolí studenta není stejné procentuální zastoupení obyvatel, co preferují střídání času.

In [329]:
p = np.sum(stridat)/np.sum(resp)
chisq(resp, stridat, p, 6)

Skutečné četnosti: [257 178 124  78  44  33  31   4]
Teoretické četnosti:  [235.58260251 162.44015169 120.89808011 104.21023939  50.4185826
  31.2453188   38.16899739   6.03602749]
Testovací kritérium:  13.058303417150736
Doplněk kritického oboru: < 0, 12.591587243743977 >


$13,058 \notin \langle 0; 12,592 \rangle$, tedy H<sub>0</sub> zamítáme.

<b>d) H<sub>0</sub>: U větších měst, menších měst a obcí je stejné procentuální zastoupení obyvatel, co preferují zimní čas.</b><br>
H<sub>A</sub>: U větších měst, menších měst a obcí není stejné procentuální zastoupení obyvatel, co preferují zimní čas.

Při ověřování následujících dvou hypotéz ( d), e) ) se pracuje se 3 skupinami, sloučenými podle větších měst, menších měst a obcí. Hodnota stupně volnosti bude tedy 1, neboť je prováděn bodový odhad (tedy 3-1-1).

In [330]:
p = np.sum(zimni_sk)/np.sum(resp_sk)
chisq(resp_sk, zimni_sk, p, 1)

Skutečné četnosti: [834 559 300]
Teoretické četnosti:  [906.97873357 512.9567503  273.06451613]
Testovací kritérium:  12.661948651569508
Doplněk kritického oboru: < 0, 3.841458820694124 >


$12,66 \notin \langle 0; 3,84 \rangle$, tedy H<sub>0</sub> zamítáme.

<b>e) H<sub>0</sub>: U větších měst, menších měst a obcí je stejné procentuální zastoupení nerozhodnutelných obyvatel.</b><br>
H<sub>A</sub>: U větších měst, menších měst a obcí není stejné procentuální zastoupení nerozhodnutelných obyvatel.

In [331]:
p = np.sum(beznazoru_sk)/np.sum(resp_sk)
chisq(resp_sk, beznazoru_sk, p, 1)

Skutečné četnosti: [337 144  57]
Teoretické četnosti:  [288.21887694 163.00692951  86.77419355]
Testovací kritérium:  20.688664757394136
Doplněk kritického oboru: < 0, 3.841458820694124 >


$20,69 \notin \langle 0; 3,84 \rangle$, tedy H<sub>0</sub> zamítáme.

f) Na základě odpovědí z okolí studenta zkuste určit z dat, zda student prováděl výzkum ve větším městě, menším městě nebo v obci. Porovnejte výsledek se skutečností a okomentujte.

In [332]:
np.set_printoptions(suppress=True) 
def chisq_f(sk_index):
    p_zimni = np.sum(zimni_sk[sk_index])/np.sum(resp_sk[sk_index])
    p_letni = np.sum(letni_sk[sk_index])/np.sum(resp_sk[sk_index])
    p_stridani = np.sum(stridat_sk[sk_index])/np.sum(resp_sk[sk_index])
    p_beznazoru = np.sum(beznazoru_sk[sk_index])/np.sum(resp_sk[sk_index])
    vyzkum = [zimni[7], letni[7], stridat[7], beznazoru[7]]
    celkem = np.sum(vyzkum)
    vyzkum_teor = [celkem*p_zimni, celkem*p_letni, celkem*p_stridani, celkem*p_beznazoru]
    print(stats.chisquare(f_obs = vyzkum, f_exp = vyzkum_teor, ddof=2))

krit_obor = stats.chi2.ppf(0.95, df=2)
print("Doplněk kritického oboru: < 0,", krit_obor, ">")
print("H0: Student prováděl výzkum ve větším městě.")
chisq_f(0)
print("H0: Student prováděl výzkum v menším městě.")
chisq_f(1)
print("H0: Student prováděl výzkum v obci.")
chisq_f(2)


Doplněk kritického oboru: < 0, 5.991464547107979 >
H0: Student prováděl výzkum ve větším městě.
Power_divergenceResult(statistic=2.438784599437811, pvalue=0.11836790712014167)
H0: Student prováděl výzkum v menším městě.
Power_divergenceResult(statistic=3.230669566985398, pvalue=0.07227113427356965)
H0: Student prováděl výzkum v obci.
Power_divergenceResult(statistic=6.947865988500661, pvalue=0.008391928065668765)


Na základě p-values lze zjistit, že hypotézy o provedení výzkumu ve větším nebo menším městě zamítnuty nebyly, naopak hypotéza o provedení výzkumu v obci byla zamítnuta. Tyto výsledky jsou srovnatelné s realitou, neboť výzkum byl prováděn v menším městě i větším městě, s převahou obyvatel z většího města. Dá se tedy vyvodit, že pozorované a dodané odpovědi souhlasí.

# Úkol 2

<b>a) Určení vhodného modelu.</b><br>
Pro vytvoření modelu a další práci s ním byla využita knihovna <i>statsmodels</i>.<br>
Daný model $Z = \beta_1 + \beta_2X + \beta_3Y + \beta_4X^2 + \beta_5Y^2 + \beta_6X*Y$ může být zjednodušen až na model $Z = \beta_1$. Je vhodné odstranit nulové parametry pro zjednodušení modelu. Ty mohou být určeny například hodnotou P>|t|, která značí p-value pro hypotézu, že daný parametr je nulový. Pokud je tedy tato hodnota vyšší, než hodnota $\alpha = 0,05$, nezamítá se hypotéza, že je parametr nulový, a může být potenciálně odstraněn. Vždy musí být přihlíženo na vhodnost modelu, která může být určena pomocí koeficientu determinace R<sup>2</sup>.

In [333]:
W = px.load_workbook('data.xlsx')
p = W['data']
dataX = np.asarray([ p['A%s'%i].value for i in range(5,75) ])
dataY = np.asarray([ p['B%s'%i].value for i in range(5,75) ])
dataZ = np.asarray([ p['D%s'%i].value for i in range(5,75) ])

F = np.column_stack((dataX, dataY, dataX**2, dataY**2, dataX*dataY))
F = sm.add_constant(F)
model = sm.OLS(dataZ, F).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.981
Model:                            OLS   Adj. R-squared:                  0.980
Method:                 Least Squares   F-statistic:                     661.5
Date:                Sun, 11 Dec 2022   Prob (F-statistic):           1.14e-53
Time:                        15:18:46   Log-Likelihood:                -242.05
No. Observations:                  70   AIC:                             496.1
Df Residuals:                      64   BIC:                             509.6
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -2.9131      3.846     -0.757      0.4

Podle hodnot P>|t| lze rozpoznat, že parametry x1, x2 a x4 lze odstranit. Hodnota R<sup>2</sup> je bez odstranění parametrů 0,981.

In [334]:
F = np.column_stack((dataX**2, dataX*dataY))
F = sm.add_constant(F)
submodel = sm.OLS(dataZ, F).fit()
print(submodel.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.980
Model:                            OLS   Adj. R-squared:                  0.979
Method:                 Least Squares   F-statistic:                     1610.
Date:                Sun, 11 Dec 2022   Prob (F-statistic):           2.29e-57
Time:                        15:18:46   Log-Likelihood:                -244.54
No. Observations:                  70   AIC:                             495.1
Df Residuals:                      67   BIC:                             501.8
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -2.4852      1.478     -1.681      0.0

Model byl postupně zjednodušen na funkci $Z = \beta_1 + \beta_4X^2 + \beta_6X*Y$. Rozdíl mezi původní a novou hodnotou R<sup>2</sup>, která je nyní 0,980, není zásadní, tedy vytvořený submodel je možno použít.

<b>b) Odhady regresních parametrů metodou nejmenších čtverců a jejich 95% intervaly spolehlivosti.</b><br>
Tyto hodnoty byly zjištěny z výše uvedeného shrnutí.
|           	| **[0.025** 	| **0.975]** 	| **Koeficient** 	|
|-----------	|------------	|------------	|----------------	|
| **const** 	|     -5.436 	|      0.466 	|        -2.4852 	|
| **x1**    	|      0.488 	|      0.525 	|         0.5065 	|
| **x2**    	|     -1.114 	|     -1.018 	|        -1.0657 	|

<b>c) Nestranně odhadněte rozptyl závislé proměnné.</b><br>
Nestranný odhad rozptylu závislé proměnné Z lze určit jako $S^2 = \frac{RSS}{n-2}$, kde RSS vyjadřuje reziduální součet čtverců. Tuto hodnotu lze z výše uvedeného modelu získat pomocí <i>mse_resid</i>, tedy $S^2 = 66,2064$.

In [335]:
submodel.mse_resid

66.20643157748601

<b>d) Vhodným testem zjistěte, že vámi zvolené dva regresní parametry jsou současně nulové.</b><br>
H<sub>0</sub>: parametry x1 = x2 = 0<br>
H<sub>A</sub>: některý z parametrů není roven 0<br>
Pro ověření hypotézy nulovosti parametrů je vhodný f-test. Hodnota $F \overset{as}{\sim} F_{1-\alpha}(k1, k2)$, k1 = I-1 a k2 = n-I, kde n je počet realizací a I je počet skupin (koeficientů). V tomto případě se jedná o hodnotu $F_{0,95}(1, 68)$. Tedy $F = 0,6249 \in \langle 0, 3,9819 \rangle$, hypotéza H<sub>0</sub> se tak na hladině významnosti 0,05 nezamítá.

In [336]:
print(model.f_test("x1=x2=0"))
print(stats.f.ppf(q=0.95, dfn=1, dfd=68))

<F test: F=0.6248911439742278, p=0.5385537138776201, df_denom=64, df_num=2>
3.9818962563017606


<b>e) Vhodným testem zjistěte, že vámi zvolené dva regresní parametry jsou současně nulové.</b><br>
H<sub>0</sub>: parametry x1 = x2<br>
H<sub>A</sub>: parametry x1 != x2<br>
Pro ověření této hypotézy byl využit t-test. Lze pozorovat, že $t \notin \langle 1,511; 1,633 \rangle$, tedy hypotéza H<sub>0</sub> se tak na hladině významnosti 0,05 zamítá.

In [339]:
print(submodel.t_test("x1=x2"))

                             Test for Constraints                             
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
c0             1.5722      0.030     51.566      0.000       1.511       1.633
